kernel/src/syscall/signalfd.rs

```diff
@@ -0,0 +1,372 @@
+                    self.pollee.invalidate();
+                }
+                None => break,
+            }
+        }
+
+        if count == 0 {
+            return_errno!(Errno::EAGAIN);
+        }
+        Ok(count * core::mem::size_of::<SignalfdSiginfo>())
+    }
+}
+
+impl Observer<SigEvents> for SignalFile {
+    fn on_events(&self, events: &SigEvents) {
+        if self
+            .signals_mask
+            .load(Ordering::Relaxed)
+            .contains(events.sig_num())
+        {
+            self.pollee.notify(IoEvents::IN);
+        }
+    }
+}
+
+impl Pollable for SignalFile {
+    fn poll(&self, mask: IoEvents, poller: Option<&mut PollHandle>) -> IoEvents {
+        self.pollee
+            .poll_with(mask, poller, || self.check_io_events())
+    }
+}
+
+impl FileLike for SignalFile {
+    fn read(&self, writer: &mut VmWriter) -> Result<usize> {
+        if writer.avail() < core::mem::size_of::<SignalfdSiginfo>() {
+            return_errno_with_message!(Errno::EINVAL, "Buffer too small for siginfo structure");
+        }
+
+        if self.is_non_blocking() {
+            self.try_read(writer)
+        } else {
+            self.wait_events(IoEvents::IN, None, || self.try_read(writer))
+        }
+    }
+
+    fn write(&self, _reader: &mut VmReader) -> Result<usize> {
+        return_errno_with_message!(Errno::EBADF, "signalfd does not support write operations");
+    }
+
+    fn status_flags(&self) -> StatusFlags {
+        if self.is_non_blocking() {
+            StatusFlags::O_NONBLOCK
+        } else {
+            StatusFlags::empty()
+        }
+    }
+
+    fn set_status_flags(&self, new_flags: StatusFlags) -> Result<()> {
+        self.set_non_blocking(new_flags.contains(StatusFlags::O_NONBLOCK));
+        Ok(())
+    }
+
+    fn metadata(&self) -> Metadata {
+        let now = RealTimeClock::get().read_time();
+        Metadata {
+            dev: 0,
+            ino: 0,
+            size: 0,
+            blk_size: 0,
+            blocks: 0,
+            atime: now,
+            mtime: now,
+            ctime: now,
+            type_: InodeType::NamedPipe,
+            mode: InodeMode::from_bits_truncate(0o400),
+            nlinks: 1,
+            uid: Uid::new_root(),
+            gid: Gid::new_root(),
+            rdev: 0,
+        }
+    }
+}
+
+impl Drop for SignalFile {
+    fn drop(&mut self) {
+        if let Some(thread) = current_thread!().as_posix_thread() {
+            thread.unregister_sigqueue_observer(&self.weak_self);
+        }
+    }
+}
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone, Pod)]
+struct SignalfdSiginfo {
+    ssi_signo: u32,
+    ssi_errno: i32,
+    ssi_code: i32,
+    ssi_pid: u32,
+    ssi_uid: u32,
+    ssi_fd: i32,
+    ssi_tid: u32,
+    ssi_band: u32,
+    ssi_overrun: u32,
+    ssi_trapno: u32,
+    ssi_status: i32,
+    ssi_int: i32,
+    ssi_ptr: u64,
+    ssi_utime: u64,
+    ssi_stime: u64,
+    ssi_addr: u64,
+    _pad: [u8; 48],
+}
+
+trait SignalToSiginfo {
```

Comment:
This trait may be better names as `ToSiganlfdSiginfo`. We already has a `siginfo` (siginfo_t), and this trait is of different meaning.